evm/chains/base.py

```diff
@@ -610,33 +612,65 @@ def validate_uncles(self, block: BaseBlock) -> None:
         """
         Validate the uncles for the given block.
         """
-        recent_ancestors = dict(
-            (ancestor.hash, ancestor)
-            for ancestor in self.get_ancestors(MAX_UNCLE_DEPTH + 1),
-        )
-        recent_uncles = []
-        for ancestor in recent_ancestors.values():
-            recent_uncles.extend([uncle.hash for uncle in ancestor.uncles])
-        recent_ancestors[block.hash] = block
-        recent_uncles.append(block.hash)
+        # Check for duplicates
+        uncle_groups = groupby(operator.attrgetter('hash'), block.uncles)
+        duplicate_uncles = tuple(sorted(
+            hash for hash, twins in uncle_groups.items() if len(twins) > 1
+        ))
+        if duplicate_uncles:
+            raise ValidationError(
+                "Block contains duplicate uncles:\n"
+                " - {0}".format(' - '.join(duplicate_uncles))
+            )
+
+        recent_ancestors = {
+            ancestor.hash: ancestor
+            for ancestor in self.get_ancestors(MAX_UNCLE_DEPTH + 1)
+        }
+        recent_uncles = {
+            uncle.hash
+            for ancestor in recent_ancestors.values()
+            for uncle in ancestor.uncles
+        }
```

Comment:
Hmm, this confused the hell out of me... it looks like you're specifying just the dict keys but no values? Maybe just something I'm not used to, or do you think it'd make sense to rewrite this without the nested comprehensions?

Reply:
It's constructing a `set` instead of a dictionary since we don't actually need the uncle here, just the hash and a check for inclusion.

I'm inclined to leave the double comprehension unless you feel strongly about it.

Reply:
Oh, set comprehensions! I'd never seen that before, and I guess that's what really confused me as I thought you were constructing a dict and I'd no idea how that was working. So yeah, just leave them nested, and thanks for the clarification!